Description: Skip everything involving mpsolve until it arrives in Debian
Author: Doug Torrance <dtorrance@piedmont.edu>
Bug-Debian: https://bugs.debian.org/958919
Last-Update: 2020-07-09

--- a/M2/configure.ac
+++ b/M2/configure.ac
@@ -670,7 +670,7 @@
 #    (Programs linked by this configure script are linked with the options in LIBS.  This allows libraries
 #    dependent on previously detected libraries to be detected by tests that involve linking.
 
-AC_SUBST(LIBLIST, " atomic_ops gc gdbm gmp mpir mpfr readline ntl flint factory lapack mpsolve frobby glpk cddlib fplll linbox gtest ")
+AC_SUBST(LIBLIST, " atomic_ops gc gdbm gmp mpir mpfr readline ntl flint factory lapack frobby glpk cddlib fplll linbox gtest ")
 # The list LIBLIST is the list of libraries that might be used and linked into M2.
 
 AC_SUBST(PROGLIST, "4ti2 gfan normaliz csdp nauty cddplus lrslib gftables topcom cohomcalg")
@@ -862,13 +862,6 @@
      fi
 fi
 
-# including mpsolve (requires gmp or mpir) 
-AC_LANG(C)
-AC_CHECK_HEADER(mps/mps.h,LIBS="-lmps $LIBS",BUILD_mpsolve=yes)
-if test $BUILD_mpsolve = yes
-then BUILTLIBS="-lmps $BUILTLIBS"
-fi
-
 AC_SUBST(LIBS_GLPK)
 AC_LANG(C)
 AC_CHECK_HEADER(glpk.h,,BUILD_glpk=yes)
--- a/M2/Macaulay2/packages/Macaulay2Doc/functions/roots-doc.m2
+++ b/M2/Macaulay2/packages/Macaulay2Doc/functions/roots-doc.m2
@@ -11,14 +11,10 @@
     EXAMPLE {
       "RR_100[x]",
       "p = x^13 + 5*x^9 + 7*x^4 + x +1",
-      "roots p",
-      "o3#0",
     },
     EXAMPLE {
       "ZZ[x]",
       "p = x^13 + 5*x^9 + 7*x^4 + x +1",
-      "roots(p^2, Precision=>150, Unique=>true)",
-      "o7#0",
     },
     PARA {
       "The roots are computed using ", TO "MPsolve", ".",
--- a/M2/Macaulay2/d/version.dd
+++ b/M2/Macaulay2/d/version.dd
@@ -19,10 +19,6 @@
    #include <mathic.h>
    #include <gdbm.h>
    #include <mathicgb.h>
-   #include <mps/version.h>
-       #ifndef MPS_MAJOR_VERSION
-       #error \"MPS_MAJOR_VERSION undefined\"
-       #endif
    #define stringize0(a) #a
    #define stringize(a) stringize0(a)
    namespace constants { extern const char* const version; } /* we have to copy this from frobby until they make it more accessible */
@@ -166,7 +162,7 @@
 	       \"not present\"
 	  #endif
 	  "),
-   "mpsolve version" => Ccode(constcharstar, "stringize(MPS_MAJOR_VERSION) \".\" stringize(MPS_MINOR_VERSION) \".\" stringize(MPS_PATCH_VERSION)" ),
+   "mpsolve version" => "not present",
    "python version" => Ccode(constcharstar,"
 	#ifdef WITH_PYTHON
 		 PY_VERSION				      
--- a/M2/Macaulay2/m2/factor.m2
+++ b/M2/Macaulay2/m2/factor.m2
@@ -160,7 +160,7 @@
 
 roots = method(Options => true);
 roots RingElement := {Precision => -1, Unique => false} >> o -> p ->
-  toList apply(rawRoots(raw p, o.Precision, o.Unique), r -> new CC from r)
+  error("roots is disabled until mpsolve arrives in Debian")
 
 -- Local Variables:
 -- compile-command: "make -C $M2BUILDDIR/Macaulay2/m2 "
--- a/M2/Macaulay2/e/polyroots.cpp
+++ b/M2/Macaulay2/e/polyroots.cpp
@@ -1,4 +1,3 @@
-#include <mps/mps.h>
 #include <stdlib.h>
 #include <stdio.h>
 #include <gmp.h>
@@ -13,144 +12,7 @@
                                           long prec,
                                           int unique)
 {
-  const Ring *R = p->get_ring();
-  const PolynomialRing *P = R->cast_to_PolynomialRing();
-  const Monoid *M = P->getMonoid();
-  if (P == 0) {
-    ERROR("expected a polynomial ring");
-    return nullptr;
-  }
-  if (P->n_vars() != 1) {
-    ERROR("expected a univariate polynomial ring");
-    return nullptr;
-  }
-  if (p->is_zero()) {
-    ERROR("expected a nonzero polynomial");
-    return nullptr;    
-  }
-  const Ring *K = P->getCoefficients();
-  int lodeg,hideg; // lowest,highest degree
-  P->degree_of_var(0,p->get_value(),lodeg,hideg);
-  if (prec == -1)
-    prec = (K->get_precision() == 0 ? 53 : K->get_precision());
-
-  
-  
-  ///////////// TODO: implement roots of a polynomial, via e.g. arb or mpsolve.
-  mps_context *s = mps_context_new ();
-  mps_monomial_poly *mps_p = mps_monomial_poly_new (s, hideg);
-  mps_context_select_algorithm(s, MPS_ALGORITHM_SECULAR_GA);
-
-  const auto ID = K->ringID();
-  for (Nterm *t = p->get_value(); t != nullptr; t = t->next) {
-    int deg;
-    M->to_expvector(t->monom, &deg); // number of vars = 1
-    if (ID==M2::ring_RR or ID==M2::ring_CC) {
-      cc_doubles_struct cc;
-      if (ID==M2::ring_RR) {
-        cc.re = t->coeff.get_double();
-        cc.im = 0;
-      } else cc = *t->coeff.get_cc_doubles();
-      mps_monomial_poly_set_coefficient_d (s, mps_p, deg, cc.re, cc.im); 
-    } else if (ID==M2::ring_RRR or ID==M2::ring_CCC) {
-      mpc_t mpc_cc;
-      mpc_init2(mpc_cc,K->get_precision());
-      if (ID==M2::ring_RRR) {
-          mpfr_get_f(mpc_cc->r,t->coeff.get_mpfr(),GMP_RNDN);
-      } else {
-        mpfr_get_f(mpc_cc->r,&t->coeff.get_cc()->re,GMP_RNDN);
-        mpfr_get_f(mpc_cc->i,&t->coeff.get_cc()->im,GMP_RNDN);
-      }
-      mps_monomial_poly_set_coefficient_f (s, mps_p, deg, mpc_cc); 
-      mpc_clear(mpc_cc);
-    } else if ((ID==M2::ring_old and K->is_ZZ()) or ID==M2::ring_QQ) {
-      mpq_t q, zero;
-      mpq_init(zero);
-      mpq_init(q);
-      if (ID==M2::ring_QQ)
-        mpq_set(q, t->coeff.get_mpq());
-      else mpq_set_z(q, t->coeff.get_mpz());
-      mps_monomial_poly_set_coefficient_q(s, mps_p, deg, q, zero); 
-      mpq_clear(q);
-      mpq_clear(zero);
-    } else {
-      ERROR("'roots' expects coefficients in ZZ, QQ, RR or CC");
-      return nullptr;
-    }
-  }
-
-  /* Set the input polynomial */
-  mps_context_set_input_poly (s, MPS_POLYNOMIAL (mps_p));
-  mps_context_set_output_prec (s, prec);
-  mps_context_set_output_goal (s, MPS_OUTPUT_GOAL_APPROXIMATE);
-
-  /* Actually solve the polynomial */
-  mps_mpsolve (s);
-
-  // result to be returned
-  engine_RawRingElementArrayOrNull result = nullptr;
-  result = getmemarraytype(engine_RawRingElementArray, hideg);
-  result->len = static_cast<int>(hideg);
-
-  ////////////////
-  if (prec <= 53) {
-    /* Allocate space to hold the results. We check only floating point results in here */
-    cplx_t *result_mps = cplx_valloc (hideg);
-
-    /* Save roots computed in the vector results */
-    mps_context_get_roots_d (s, &result_mps, nullptr);
-
-    /* Print out roots */
-    // for (int i = 0; i < hideg; i++) {
-    //   cplx_out_str (stdout, result_mps[i]);
-    //   printf ("\n");
-    // }
-
-    ///////////////////////////////
-    // copy to mps_result to result
-
-    const Ring *C = IM2_Ring_CCC(prec);
-    for (int i = 0; i < hideg; i++) {
-      auto& mps_root = result_mps[i];
-      ring_elem m2_root;
-      C->from_complex_double(cplx_Re(mps_root), cplx_Im(mps_root),
-                             m2_root);
-      result->array[i] = RingElement::make_raw(C, m2_root);
-    }
- 
-    free (result_mps);
-  } else { // prec > 53
-    mpc_t *roots = nullptr;
-    rdpe_t *radii = nullptr;
-
-    mps_context_get_roots_m (s, &roots, &radii);
-    /* Sort roots in the order of increasing real part */
-    //sort_roots(hideg, roots, radii);
-    // for (int i = 0; i < hideg; i++) {
-    //   mpc_out_str (stdout, 10, prec/4, roots[i]);
-    //   printf ("\n");
-    // }
-    
-    const Ring *C = IM2_Ring_CCC(prec);
-    const M2::ARingCCC C0(prec);
-
-    M2::ARingCCC::ElementType cc;
-    C0.init(cc);
-    for (int i = 0; i < hideg; i++) {
-      auto& mps_root = roots[i];
-      mpfr_set_f(&cc.re,mpc_Re(mps_root),GMP_RNDN);
-      mpfr_set_f(&cc.im,mpc_Im(mps_root),GMP_RNDN);
-      ring_elem m2_root;
-      C0.to_ring_elem(m2_root, cc);
-      result->array[i] = RingElement::make_raw(C, m2_root);
-    }
-    C0.clear(cc);
-        
-    free(roots);
-    free(radii);
-  }
-
-  return result;
+  return nullptr;
 }
 
 // Local Variables:
